File: 0078-subsets/0078-subsets.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums, i = 0) {
  if (i === nums.length) return [[]];

  const first = nums[i];
  const subsetsWithoutFirst = subsets(nums, i + 1);
  const subsetsWithFirst = subsetsWithoutFirst.map((sub) => [first, ...sub]);

  return [...subsetsWithoutFirst, ...subsetsWithFirst];
};

File: 0078-subsets/0078-subsets.py
class Solution:
    def subsets(self, nums):
      if len(nums) == 0:
        return [[]]
      
      first = nums[0]
      sets_without_first = self.subsets(nums[1:])
      sets_with_first = []
      for combo in sets_without_first:
        sets_with_first.append([ first, *combo ])
      
      return sets_with_first + sets_without_first
        